### ISSUE TITLE
[rllib] Learning rate schedule is ignored

### ISSUE THREAD

**Issue by gordoni** [ray-project/ray#6096]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Ray installed from (source or binary)**: binary
- **Ray version**: tested on 0.7.3 but the bug has existed since 0.7.1 and still exists in the latest 0.7.6+ sources
- **Python version**: 3.6.5
- **Exact command to reproduce**:

Edit tf_policy.py to report use of LearningRateSchedule mixin:

```
diff --git a/ray/python/ray/rllib/policy/tf_policy.py b/ray/python/ray/rllib/policy/tf_policy.py
index 160fc9c..71c72b9 100644
--- a/ray/python/ray/rllib/policy/tf_policy.py
+++ b/ray/python/ray/rllib/policy/tf_policy.py
@@ -571,6 +571,7 @@ class LearningRateSchedule(object):
 
     @override(TFPolicy)
     def optimizer(self):
+        print('Using LearningRateSchedule mixin optimizer.') # Never gets executed!
         return tf.train.AdamOptimizer(self.cur_lr)
 
 ```

Now train an agent that makes use of the LearningRateSchedule mixin (such as PPO):

```
./train.py --run=PPO --env=MountainCarContinuous-v0 \
    --config='{"lr_schedule": [[0, 1e-5], [10000000, 1e-6]], "num_workers": 1}'
```

And the above added print statement never gets executed!

### Describe the problem

The learning rate schedule is not being used, as shown by the LearningRateSchedule mixin optimizer() never being called.

The problem appears due to tf_policy_template.py calling a hard coded TFPolicy optimizer() rather than the mixin version of the optimizer() as illustrated by the following patch (which fixes the problem for me):

```
diff --git a/ray/python/ray/rllib/policy/tf_policy_template.py b/ray/python/ray/rllib/policy/tf_policy_template.py
index e44a6da..6ff2f59 100644
--- a/ray/python/ray/rllib/policy/tf_policy_template.py
+++ b/ray/python/ray/rllib/policy/tf_policy_template.py
@@ -161,7 +161,7 @@ def build_tf_policy(name,
             if optimizer_fn:
                 return optimizer_fn(self, self.config)
             else:
-                return TFPolicy.optimizer(self)
+                return base.optimizer(self)
 
         @override(TFPolicy)
         def gradients(self, optimizer, loss):
```

This patch hasn't been tested extensively, and there are additional locations in tf_policy_template.py where TFPolicy is hard coded which may need to be replaced for other mixins.

**Comment by ericl** [comment]
Great find! Could you open a PR with this fix?

**Comment by ericl** [comment]
Actually try https://github.com/ray-project/ray/pull/6101 out ?

**Comment by gordoni** [comment]
#6101 works for me, with the caveat that I tested the patch out on a slightly modified version of Ray 0.7.3. Thanks!
